sqldb/invoices.go

```diff
@@ -0,0 +1,1651 @@
+	*invpkg.Invoice, error) {
+
+	var updatedInvoice *invpkg.Invoice
+
+	txOpt := InvoiceQueriesTxOptions{readOnly: false}
+	txErr := i.db.ExecTx(ctx, &txOpt, func(db InvoiceQueries) error {
+		invoice, err := i.fetchInvoice(ctx, db, ref)
+		if err != nil {
+			return err
+		}
+
+		updateTime := i.clock.Now()
+		updater := &sqlInvoiceUpdater{
+			db:         db,
+			ctx:        ctx,
+			invoice:    invoice,
+			updateTime: updateTime,
+		}
+
+		payHash := ref.PayHash()
+		updatedInvoice, err = invpkg.UpdateInvoice(
+			payHash, invoice, updateTime, callback, updater,
+		)
+
+		return err
+	})
+	if txErr != nil {
+		return updatedInvoice, txErr
+	}
+
+	return updatedInvoice, nil
+}
+
+// DeleteInvoice attempts to delete the passed invoices and all their related
+// data from the database in one transaction.
+//
+// NOTE: Settled invoices cannot be deleted with this method.
+func (i *InvoiceStore) DeleteInvoice(ctx context.Context,
+	invoicesToDelete []invpkg.InvoiceDeleteRef) error {
+
+	// All the InvoiceDeleteRef instances include the add index of the
+	// invoice. The rest was added to ensure that the invoices were deleted
+	// properly in the kv database. When we have fully migrated we can
+	// remove the rest of the fields.
+	for _, ref := range invoicesToDelete {
+		if ref.AddIndex == 0 {
+			return fmt.Errorf("unable to delete invoice using a "+
+				"ref without AddIndex set: %v", ref)
+		}
+	}
+
+	var writeTxOpt InvoiceQueriesTxOptions
+	err := i.db.ExecTx(ctx, &writeTxOpt, func(db InvoiceQueries) error {
+		for _, ref := range invoicesToDelete {
+			params := sqlc.DeleteInvoiceParams{
+				AddIndex: sqlInt64(ref.AddIndex),
+			}
+
+			if ref.SettleIndex != 0 {
+				params.SettledIndex = sqlInt64(ref.SettleIndex)
+			}
+
+			if ref.PayHash != lntypes.ZeroHash {
+				params.Hash = ref.PayHash[:]
+			}
+
+			result, err := db.DeleteInvoice(ctx, params)
+			if err != nil {
+				return fmt.Errorf("unable to delete "+
+					"invoice(%v): %w", ref.AddIndex, err)
+			}
+			rowsAffected, err := result.RowsAffected()
+			if err != nil {
+				return fmt.Errorf("unable to get rows "+
+					"affected: %w", err)
+			}
+			if rowsAffected == 0 {
+				return fmt.Errorf("%w: %v",
+					invpkg.ErrInvoiceNotFound, ref.AddIndex)
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("unable to delete invoices: %w", err)
+	}
+
+	return nil
+}
+
+// DeleteCanceledInvoices removes all canceled invoices from the database.
+func (i *InvoiceStore) DeleteCanceledInvoices(ctx context.Context) error {
+	var writeTxOpt InvoiceQueriesTxOptions
+	err := i.db.ExecTx(ctx, &writeTxOpt, func(db InvoiceQueries) error {
+		_, err := db.DeleteCanceledInvoices(ctx)
+		if err != nil {
+			return fmt.Errorf("unable to delete canceled "+
+				"invoices: %w", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("unable to delete invoices: %w", err)
+	}
+
+	return nil
+}
+
+// fetchInvoiceData fetches the common invoice data for the given params.
```

Comment:
I would prefer a way more detailed comment,  because it does not really fetch the invoice data because it already gets the `sqlc.Invoice` but instead all the additional tables which tie to the invoice table. So I think its worth to be more specific here.

Reply:
PTAL